File: client/src/types/garment.ts
/**
 * Garment types and measurement schemas.
 *
 * Mirrors the server-side GarmentMeasurements model.
 */

export type GarmentType = "tshirt" | "pants" | "dress";

/** Human-readable labels for garment types. */
export const GARMENT_TYPE_LABELS: Record<GarmentType, string> = {
  tshirt: "T-Shirt",
  pants: "Pants",
  dress: "Dress",
};

/** Body measurements (for conforming garment from body mesh). */
export type BodyMeasurementsForGarment = Record<string, number | string | boolean>;

/** Measurements for sizing a 3D garment. */
export interface GarmentMeasurements {
  garment_type: GarmentType;
  chest_cm?: number;
  waist_cm?: number;
  hip_cm?: number;
  length_cm?: number;
  sleeve_length_cm?: number;
  inseam_cm?: number;
  /** Body landmarks from SMPL (optional). */
  body_landmarks?: Record<string, number>;
  /** Body measurements used to generate the model (optional). When present, garment is built as conforming offset. */
  body_measurements?: BodyMeasurementsForGarment;
}

/** Options when submitting the garment upload form (reserved for future use). */
export interface GarmentSubmitOptions {
  [key: string]: unknown;
}

/** A single garment loaded into the try-on scene. */
export interface GarmentItem {
  /** Unique identifier for this garment instance. */
  id: string;
  /** Garment type. */
  type: GarmentType;
  /** Object URL pointing to the loaded GLB blob. */
  modelUrl: string;
  /** Measurements used to generate this garment. */
  measurements: GarmentMeasurements;
  /** Whether the garment is visible in the scene. */
  visible: boolean;
  /** Original filename of the uploaded image. */
  imageName: string;
}

/** Default measurements per garment type. */
export const DEFAULT_GARMENT_MEASUREMENTS: Record<GarmentType, GarmentMeasurements> = {
  tshirt: {
    garment_type: "tshirt",
    chest_cm: 100,
    length_cm: 72,
    sleeve_length_cm: 24,
  },
  pants: {
    garment_type: "pants",
    waist_cm: 84,
    hip_cm: 100,
    inseam_cm: 78,
    length_cm: 100,
  },
  dress: {
    garment_type: "dress",
    chest_cm: 92,
    waist_cm: 74,
    hip_cm: 98,
    length_cm: 100,
  },
};

File: server/app/models/garment.py
"""Garment data models.

Defines the schema for garment types and measurements used
when converting a 2D outfit image into a 3D garment mesh.
"""

from enum import Enum

from pydantic import BaseModel, Field


class GarmentType(str, Enum):
    """Supported garment template types."""

    TSHIRT = "tshirt"
    POLO = "polo"
    BUTTON_DOWN = "button_down"
    HOODIE = "hoodie"
    JACKET = "jacket"
    PANTS = "pants"
    DRESS = "dress"


class GarmentMeasurements(BaseModel):
    """Measurements for sizing a 3D garment.

    Not all fields are required for every garment type.
    Provide the fields relevant to the selected garment_type.
    """

    garment_type: GarmentType = Field(
        description="Type of garment template to use",
    )

    # -- Shared measurements --------------------------------------------------
    length_cm: float | None = Field(
        default=None,
        ge=20.0,
        le=150.0,
        description="Garment length in centimeters",
    )

    # -- Top / dress measurements ---------------------------------------------
    chest_cm: float | None = Field(
        default=None,
        ge=60.0,
        le=160.0,
        description="Chest circumference (tops and dresses)",
    )
    sleeve_length_cm: float | None = Field(
        default=None,
        ge=10.0,
        le=90.0,
        description="Sleeve length (tops only)",
    )

    # -- Bottom / dress measurements ------------------------------------------
    waist_cm: float | None = Field(
        default=None,
        ge=50.0,
        le=150.0,
        description="Waist circumference (pants and dresses)",
    )
    hip_cm: float | None = Field(
        default=None,
        ge=60.0,
        le=160.0,
        description="Hip circumference (pants and dresses)",
    )
    inseam_cm: float | None = Field(
        default=None,
        ge=20.0,
        le=100.0,
        description="Inseam length (pants only)",
    )

    # -- Body landmarks (optional, from SMPL body generation) --------------------
    body_landmarks: dict | None = Field(
        default=None,
        description="Body landmark positions from SMPL body generation (auto-populated by frontend)",
    )
    # -- Body measurements (optional, for conforming garment from body mesh) ------
    body_measurements: dict | None = Field(
        default=None,
        description="Body measurements used to generate the body; when provided with body_landmarks, garment is built as conforming offset from body mesh",
    )

    model_config = {"json_schema_extra": {"examples": [
        {
            "garment_type": "tshirt",
            "chest_cm": 100,
            "length_cm": 72,
            "sleeve_length_cm": 24,
        }
    ]}}
